### PR TITLE
chore: made Rust CI workflow

### PR DESCRIPTION
**Pull request by taeh98**
- This PR adds CI for your Rust code
- It runs multiple checks on your code automatically to check it for errors
- It checks it compiles, runs any tests you've made, runs lints, and checks for security vulnerabilities
- This helps standardise your code and detect issues earlier and more easily, saving time and effort

### CODE REVIEW

File: .github/workflows/rust_ci.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      - name: Run cargo check
+        run: cargo check
+
+  test:
+    name: Run tests
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: dtolnay/rust-toolchain@stable
+      - uses: Swatinem/rust-cache@v2
```
Comment (Spydr06) [inline .github/workflows/rust_ci.yml:38]: Two things with this:
- LogicRs requires the nightly toolchain (idk if that's available with this toolchain provider)
- I'd prefer if we used the official rust solution and not depend on third-party tools. I'm not _that_ into CI, so if there's a good reason against that, please lmk

**Review by Spydr06 — CHANGES_REQUESTED** [review]
Thanks for this and great idea! I've added two things that need changing below

File: .github/workflows/rust_ci.yml (hunk 1 above)
Comment (taeh98) [reply .github/workflows/rust_ci.yml:38]: Hey, thanks for your feedback.

To your first point, yeah, I'll change it over to the nightly toolchain if that's what your code needs.

To your second, [there isn't an official Rust GitHub action, but dtolnay/rust-toolchain is now what most people use](https://github.com/rust-lang/rustup/issues/3409). I think the checkout action has been updated though, so I'll update it and also recheck everything else is up-to-date too.

Comment (taeh98) [reply .github/workflows/rust_ci.yml]: OK @Spydr06,

I just finished moving all the actions over to the nightly toolchain, updating the checkout action used, and moving everything over to the de-facto standard actions used for jobs. Please let me know if you have any further thoughts. Thanks! :)

Comment (Spydr06) [reply .github/workflows/rust_ci.yml]: > To your second, https://github.com/rust-lang/rustup/issues/3409. I think the checkout action has been updated though, so I'll update it and also recheck everything else is up-to-date too.

Ah ok, did not know that. Then it's fine ofc

**Review by Spydr06 — APPROVED** [review]
Looks good, I hope this works :P

**Comment by Spydr06** [comment]
image
some system libs are missing, can you add the following packages and their `-dev` versions to the container?

- `gtk4`, `gtk4-dev`
- `libcairo`, `libcairo-dev`
- `libadwaita`, `libadwaita-dev`

**Comment by taeh98** [comment]
> @taeh98 image some system libs are missing, can you add the following packages and their `-dev` versions to the container?
> 
>     * `gtk4`, `gtk4-dev`
> 
>     * `libcairo`, `libcairo-dev`
> 
>     * `libadwaita`, `libadwaita-dev`

Hey @Spydr06, sorry it isn't working :/

Are those apt dependencies (or something similar)? The jobs run in an Ubtuntu container, so I imagine fixing this issue would just require running an extra apt install step if so? I'll make a new branch to fix them since you merged this MR.

**Comment by Spydr06** [comment]
yep exactly, you need to install them using `apt`. It should be the following packages:
`build-essential`, `libadwaita-1-dev`, `libcairo2-dev`, `libgtk-4-dev`

**Comment by taeh98** [comment]
@Spydr06 looking at the output of [this job](https://github.com/taeh98/logicrs/actions/runs/6372522524/job/17295370440), are you sure the dependencies are right? I'm currently installing:
`packages: build-essential libadwaita-1-dev libcairo2-dev libgtk-4-dev`

**Comment by Spydr06** [comment]
should be correct, other stuff like `libpango` (from the screenshot) is a dependency iirc.

**Comment by Spydr06** [comment]
if it doesn't work, we can always add it later since the CI will now trigger on each commit automatically

**Comment by taeh98** [comment]
@Spydr06 I don't think they're the right dependencies. Looking at [this commit](https://github.com/taeh98/logicrs/commit/1a2fa936e749b69f4e4c8e1c27b05403d2b713d8) where I added the dependencies you mentioned, [the CI checks still fail because of missing dependencies](https://github.com/taeh98/logicrs/actions/runs/6372522524/job/17295370440). Are the dependencies definitely right and all you need to run your code? Is there an extra step required, like a build step maybe?

Here's the error:

```
error: failed to run custom build command for `libadwaita-sys v0.2.1`

Caused by:
  process didn't exit successfully: `/home/runner/work/logicrs/logicrs/target/debug/build/libadwaita-sys-2f78e7a6d0c10bf9/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=LIBADWAITA_1_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:warning=`PKG_CONFIG_ALLOW_SYSTEM_CFLAGS="1" "pkg-config" "--libs" "--cflags" "libadwaita-1" "libadwaita-1 >= 1.2"` did not exit successfully: exit status: 1
  error: could not find system library 'libadwaita-1' required by the 'libadwaita-sys' crate

  --- stderr
  Requested 'libadwaita-1 >= 1.2' but version of Adwaita is 1.1.7

warning: build failed, waiting for other jobs to finish...
Error: Process completed with exit code 101.
```

**Comment by Spydr06** [comment]
It seems like the version of `libadwaita` Ubunu ships with is outdated...

**Comment by Spydr06** [comment]
I can try if there's a newer package locally

**Comment by taeh98** [comment]
> I can try if there's a newer package locally

OK, let me know what happens please :)

**Comment by Spydr06** [comment]
Ok, it compiles without problem on ubuntu 23.04, your current CI setup uses 22.10 (because it's LTS). Try changing the `ubuntu-latest` to `ubuntu-23.04` or sth

**Pull request merged**